### PR TITLE
[cuDNN][SDPA] Declare the attn_mask dtype to cuDNN instead of inheriting the graph io dtype

### PR DESCRIPTION
**Pull request by YangXu1990uiuc**
## The bug

`scaled_dot_product_attention` accepts an `attn_mask` whose dtype differs from `query` —
`validate_sdpa_input` allows `bool`, `float32`, or `query.dtype`. The cuDNN backend does not honour
that: `aten/src/ATen/native/cudnn/MHA.cpp` derives one graph-wide io data type from `query`, and
creates the bias tensor with dim and stride but no data type of its own, so an fp32 mask is
described to cuDNN as fp16/bf16 and reinterpreted element by element. Output and gradients are
silently wrong — no exception, no warning.

This is the default path on SM 9.0 / 10.x, where `check_prefer_cudnn_attention()` puts cuDNN first
when the runtime cuDNN is newer than 9.15.

It takes no unusual code to hit. `nn.Transformer.generate_square_subsequent_mask` returns float32
regardless of the module's dtype:

```python
S, E, H = 32, 64, 4
mask = nn.Transformer.generate_square_subsequent_mask(S, device="cuda")   # torch.float32
mha = nn.MultiheadAttention(E, H, batch_first=True).cuda().half().eval()
x = torch.randn(2, S, E, device="cuda", dtype=torch.float16)
out, _ = mha(x, x, x, attn_mask=mask, need_weights=False)
torch.isnan(out).float().mean()      # 0.969
```

The memory-efficient backend rejects the same input loudly
(`invalid dtype for bias - should match query's dtype`); cuDNN accepts it and misreads it.

With finite mask values rather than the `-inf` of a causal mask, the output is finite but
numerically wrong instead of NaN, which is easier to miss.

## The fix

Declare the bias with its actual dtype at the two live construction sites, forward and backward.
cuDNN supports a float bias with half io, so nothing is cast — keeping the mask in fp32 avoids both
an extra device-side copy of the largest tensor in the graph and the precision loss of folding
large-magnitude fp32 mask values into half.

The bias dtype also has to enter the execution-plan cache key. `MHAParams` records the bias sizes
and strides only and the cache hashes the struct byte-for-byte, so two calls differing only in mask
dtype currently share one plan; whichever dtype the process saw first wins, which is why this can
present as intermittent. `MHAParams` is `memset` to zero before use, so the no-bias case is a
deterministic value whatever enumerator happens to be zero, and the hash is stable. The new field
sits immediately after the existing `fe::DataType_t dataType`, so the struct grows by exactly four
bytes with no new padding hole — which matters because `ParamsHash` reads `sizeof(MHAParams)` bytes
raw.

## Evidence

A/B on one build — same commit, same cuDNN 9.26, only `MHA.cpp` differs:

| | forward/backward matrix (12 cases) | new tests, SM 10.0 | new tests, SM 9.0 |
|---|---|---|---|
| `origin/main` | **9 fail** | **3 failed** | **3 failed** |
| this PR | **0 fail** | **3 passed** | **3 passed** |

<details>
<summary>The failing cases on <code>origin/main</code>, output compared against the MATH backend in fp32</summary>

```
  fwd  qkv=float16  mask=float16    finite=True   max_abs_err=5.0e-04    OK    <- control
  fwd  qkv=float16  mask=float32    finite=False  max_abs_err=nan        FAIL
  fwd  qkv=bfloat16 mask=bfloat16   finite=True   max_abs_err=3.9e-03    OK    <- control
  fwd  qkv=bfloat16 mask=float32    finite=False  max_abs_err=nan        FAIL

  bwd  qkv=float16  mask=float32  dQ/dK/dV        all non-finite         FAIL
  bwd  qkv=bfloat16 mask=float32  dQ/dK/dV        all non-finite         FAIL

  cache  same-dtype mask first                    finite=True            OK
  cache  then fp32 mask, identical shape          finite=False           FAIL
```

The matched-dtype rows are controls and are unaffected either way. With finite mask values the
forward rows give max abs diff 2.986 over 99.1% of elements instead of NaN.
</details>

## Tests

Three tests in `test/test_transformers.py`, beside the existing cuDNN attention coverage: forward
across fp16/bf16 with matched and fp32 masks (matched cases are the control), backward checking
dQ/dK/dV, and one that issues both mask dtypes at the same shape in one process, which fails
without the cache-key change.

## Why existing tests did not catch it

The cuDNN mask tests in `test_transformers.py` use either bool masks — which
`native/transformers/attention.cpp` canonicalizes to an additive mask in `query`'s dtype before
dispatch, so they never reach cuDNN as bool — or floating masks already matching `query`'s dtype
(`torch.zeros_like(attn_mask, dtype=query.dtype)`). None covers an fp32 additive mask against
fp16/bf16 q, k, v.

## Not included

The backward graph also bakes `grad_output`'s layout into the cached plan without recording it in
the key. Different root cause, separate reproducer, separate PR; this one stays scoped to the bias
dtype.

## Caveats

- Verified on SM 9.0 and SM 10.0. The build used here targets those two architectures only, so the
  fix has not been *executed* on SM 8.0/8.9; on those the failure reproduces on `main` but cuDNN is
  not the preferred backend. SM 12.x was not available.
- No performance measurement yet. An fp32 bias may select a different engine and does read more
  bias data than a half one; that is a performance question to measure, not a reason to keep an
  incorrect dtype declaration.
- `bias_data_type` rejects dtypes other than half/bfloat16/float. Bool cannot reach it through
  `F.scaled_dot_product_attention`; a direct call to the raw ATen op can still pass one and will now
  get a clear error instead of a silently misread buffer.

## AI assistance disclosure

Per [`AI_POLICY.md`](https://github.com/pytorch/pytorch/blob/main/AI_POLICY.md): this change was
prepared with AI assistance (Claude Code). Concretely, the agent was used to investigate the cuDNN
SDPA integration layer, to draft the diff and the tests, and to run the measurements — the A/B table
and the failing-case matrix above are its output, reproduced on the hardware named beside them.

The parts that were not the agent's:

- The investigation turned up a second claim, that the cuDNN dropout guard needlessly rejects
  `dropout_p = 0.1`. That is wrong and was discarded: cuDNN's SDPA dropout uses a 4-bit RNG, so the
  realised probability is quantised to multiples of 1/16 and the guard is correct. The tool had only
  measured that the call succeeds, not the realised drop rate. Mentioning it because it is the
  reason I do not treat the rest of the output as self-certifying.
- An independent review pass before submitting produced two corrections that are already folded in:
  the original text misidentified why the existing tests miss this (bool masks are canonicalised to
  `query.dtype` in `native/transformers/attention.cpp`, so they never reach cuDNN as bool), and it
  asserted the fix "costs nothing in performance", which the evidence does not support — that claim
  is now stated as an open measurement in Caveats.

I have read the diff and the surrounding code, understand the change, and take responsibility for
it. Happy to walk through any part of the reasoning that would be easier to check in review than in
prose.

### CODE REVIEW

File: test/test_transformers.py [hunk 1]
```diff
@@ -3913,6 +3913,57 @@ def test_cudnn_attention_mask_broken_177842(self):
             )
         self.assertEqual(attn_output_math, attn_output_cudnn, atol=5e-3, rtol=3e-3)
 
+    @unittest.skipIf(not PLATFORM_SUPPORTS_CUDNN_ATTENTION, "cudnn Attention is not supported on this system")
```
Comment (eqy) [inline test/test_transformers.py:3916]: These tests look very similar and only seem to parameterize on `dtype`? Can they be folded into a single unit test definition with more parameterization?

File: test/test_transformers.py [hunk 2]
```diff
@@ -3913,6 +3913,57 @@ def test_cudnn_attention_mask_broken_177842(self):
...
+        grad_out = torch.randn_like(out)
+        grads = torch.autograd.grad(out, (query, key, value), grad_out)
+        grads_ref = torch.autograd.grad(ref, (query, key, value), grad_out)
+        for grad, grad_ref in zip(grads, grads_ref):
+            self.assertTrue(grad.isfinite().all())
+            self.assertEqual(grad, grad_ref, atol=4 * atol, rtol=4 * rtol)
+
+    @unittest.skipIf(not PLATFORM_SUPPORTS_CUDNN_ATTENTION, "cudnn Attention is not supported on this system")
+    def test_cudnn_attention_attn_mask_dtype_cache_key(self, device):
+        # The graph cache is keyed on the bias sizes and strides only, so masks that differ
```
Comment (eqy) [inline test/test_transformers.py]: Doesn't this PR explicitly add the bias/mask dtype to the key, so the comment seems misleading?

**Review by eqy — APPROVED** [review]

**Comment by eqy** [comment]
@claude review this PR, pay attention to any subtle correctness issues that would be difficult for humans to catch and regressions that could be introduced

File: test/test_transformers.py (hunk 1 above)
Comment (YangXu1990uiuc) [reply test/test_transformers.py:3916]: Folded into one definition parametrized on `(dtype, mask_dtype)` in f1fe4ba. The alternating-dtype loop stays inside the body rather than becoming a third parametrization axis, because the cache-key half depends on both dtypes being issued in the same process at the same shape — as separate variants that would silently depend on the order pytest runs them in.

File: test/test_transformers.py (hunk 2 above)
Comment (YangXu1990uiuc) [reply test/test_transformers.py]: You are right, that comment described the pre-fix behaviour in the present tense and read as if the dtype were still missing from the key. Reworded in f1fe4ba to say that the two masks shared one cached plan *before* the bias dtype was added to the key, which is what the test is guarding against.

**Comment by YangXu1990uiuc** [comment]
Thanks — all three smaller points are in `5c40035`:

- `bias_data_type` is now `static`. You are right that it sat above the anonymous namespace; the neighbouring helpers having the same issue is not a reason to add another.
- the matched-dtype variant aliased `other` to `mask`, so it ran four identical forwards and compared a tensor's strides with its own. It now runs once as a control, and the alternating loop is kept only for the case where the dtypes actually differ.
- added the note about `autograd.grad` being outside the `sdpa_kernel` context on purpose.

Also corrected the PR description: you are right that the `NOT_SET == 0` argument was doing work it did not need to. What matters is only that `memset` makes the no-bias case deterministic, and `has_attn_bias` plus the bias dim/stride would separate the keys regardless. Replaced it with the padding argument, which is the part that is actually load-bearing.

On the main concern — agreed, and I would rather not reason about it. The unverified half of the matrix is exactly SM 8.x with whatever cuDNN the CI images ship, and `PLATFORM_SUPPORTS_CUDNN_ATTENTION` is `SM80OrLater && cuDNN >= 9.0`, so the new test does run there. I am treating the cuDNN-attention shards on this PR as the evidence rather than the local SM 9.0/10.0 runs, and will report what they say before asking for a merge. If they fail on an older cuDNN, a bias-dtype gate in `can_use_cudnn_attention` is the right shape — a cast would reintroduce the precision loss this is avoiding.

The nested-tensor cache observation is a real bug and matches what I saw separately: both nested paths look the key up and never insert, so the varlen graph is rebuilt every call. Out of scope here, filing separately.

**Comment by eqy** [comment]
nested tensor issue should be resolved by https://github.com/pytorch/pytorch/pull/191576
